File: routers/webhooks.py
import httpx

from fastapi import (
    APIRouter,
    HTTPException,
    Request,
)
from fastapi.responses import Response

from core.config import settings
from core.security import (
    verify_workflow_token,
)
from db.database import get_db


router = APIRouter(
    prefix="/webhooks",
    tags=["Webhooks"],
)


HOP_BY_HOP_HEADERS = {
    "connection",
    "keep-alive",
    "proxy-authenticate",
    "proxy-authorization",
    "te",
    "trailer",
    "transfer-encoding",
    "upgrade",
}


def ensure_active_workflow(
    client_id: str,
    workflow_id: str,
) -> None:

    conn = get_db()

    row = conn.execute(
        """
        SELECT workflow_id
        FROM client_workflows
        WHERE client_id = ?
          AND workflow_id = ?
          AND status = 'active'
        """,
        (
            client_id,
            workflow_id,
        ),
    ).fetchone()

    conn.close()

    if not row:
        raise HTTPException(
            status_code=403,
            detail="Workflow is not authorized",
        )


@router.api_route(
    "/{workflow_id}",
    methods=[
        "POST",
        "PUT",
        "PATCH",
        "DELETE",
    ],
)
async def gateway(
    workflow_id: str,
    request: Request,
):

    authorization = request.headers.get(
        "Authorization",
        "",
    )

    if not authorization.startswith(
        "Bearer "
    ):
        raise HTTPException(
            status_code=401,
            detail="Unauthorized",
        )

    token = authorization[7:].strip()

    claims = verify_workflow_token(
        token,
        workflow_id,
    )

    client_id = claims["sub"]

    ensure_active_workflow(
        client_id,
        workflow_id,
    )

    body = await request.body()

    headers = {}

    for key, value in request.headers.items():

        lower = key.lower()

        if lower in HOP_BY_HOP_HEADERS:
            continue

        if lower in {
            "authorization",
            "host",
            "content-length",
        }:
            continue

        headers[key] = value

    # These headers are trusted because
    # the caller cannot supply the originals.
    headers[
        "X-T3Z-Client-Id"
    ] = client_id

    headers[
        "X-T3Z-Workflow-Id"
    ] = workflow_id

    target_url = (
    f"{settings.n8n_base_url}/"
    f"{settings.n8n_webhook_prefix}/"
    f"{workflow['webhook_path']}"
)

    try:
        async with httpx.AsyncClient(
            timeout=30.0
        ) as client:

            upstream = await client.request(
                method=request.method,
                url=target_url,
                params=request.query_params,
                headers=headers,
                content=body,
            )

    except httpx.RequestError as exc:
        raise HTTPException(
            status_code=502,
            detail="n8n upstream unavailable",
        ) from exc

    response_headers = {
        key: value
        for key, value in upstream.headers.items()
        if key.lower()
        not in HOP_BY_HOP_HEADERS
    }

    return Response(
        content=upstream.content,
        status_code=upstream.status_code,
        headers=response_headers,
        media_type=upstream.headers.get(
            "content-type"
        ),
    )

    def get_workflow(
        client_id: str,
        workflow_id: str,
    ) -> dict:

        conn = get_db()

        row = conn.execute(
            """
            SELECT
                client_id,
                workflow_id,
                n8n_workflow_id,
                webhook_path,
                status
            FROM client_workflows
            WHERE client_id = ?
            AND workflow_id = ?
            AND status = 'active'
            """,
            (
                client_id,
                workflow_id,
            ),
        ).fetchone()

        conn.close()

        if not row:
            raise HTTPException(
                status_code=403,
                detail="Workflow not authorized",
            )

        return dict(row)

File: internal/handlers/webhooks.go
package handlers

import (
	"bytes"
	"database/sql"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"strings"

	"github.com/go-chi/chi/v5"
	"t3z/api-gateway/internal/database"
	"t3z/api-gateway/internal/models"
	"t3z/api-gateway/internal/security"
	"t3z/api-gateway/internal/services"
)

var hopByHopHeaders = map[string]bool{
	"connection":          true,
	"keep-alive":          true,
	"proxy-authenticate":  true,
	"proxy-authorization": true,
	"te":                  true,
	"trailer":             true,
	"transfer-encoding":   true,
	"upgrade":             true,
	"authorization":       true,
	"host":                true,
	"content-length":      true,
}

type WebhooksHandler struct {
	db       *database.DB
	jwt      *security.JWTService
	windmill *services.WindmillService
}

func NewWebhooksHandler(db *database.DB, jwt *security.JWTService, wm *services.WindmillService) *WebhooksHandler {
	return &WebhooksHandler{
		db:       db,
		jwt:      jwt,
		windmill: wm,
	}
}

func (h *WebhooksHandler) Gateway(w http.ResponseWriter, r *http.Request) {
	workflowID := chi.URLParam(r, "workflow_id")
	if workflowID == "" {
		respondJSON(w, http.StatusBadRequest, models.ErrorResponse{Detail: "Missing workflow_id"})
		return
	}

	authHeader := r.Header.Get("Authorization")
	if !strings.HasPrefix(authHeader, "Bearer ") {
		respondJSON(w, http.StatusUnauthorized, models.ErrorResponse{Detail: "Unauthorized"})
		return
	}

	token := strings.TrimSpace(authHeader[7:])
	claims, err := h.jwt.VerifyWorkflowToken(token, workflowID)
	if err != nil {
		status := http.StatusUnauthorized
		if strings.Contains(err.Error(), "not authorized") {
			status = http.StatusForbidden
		}
		respondJSON(w, status, models.ErrorResponse{Detail: err.Error()})
		return
	}

	clientID := claims.Subject

	// Verify active workflow in database and retrieve Windmill path
	var windmillPath, webhookPath sql.NullString
	err = h.db.QueryRow(`
		SELECT windmill_path, webhook_path
		FROM client_workflows
		WHERE client_id = ? AND workflow_id = ? AND status = 'active'
	`, clientID, workflowID).Scan(&windmillPath, &webhookPath)

	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			respondJSON(w, http.StatusForbidden, models.ErrorResponse{Detail: "Workflow is not authorized"})
			return
		}
		respondJSON(w, http.StatusInternalServerError, models.ErrorResponse{Detail: "Database error"})
		return
	}

	targetPath := windmillPath.String
	if targetPath == "" {
		targetPath = webhookPath.String
	}
	if targetPath == "" {
		targetPath = fmtPath(clientID, workflowID)
	}

	// Create short-lived integration token for Google Sheets proxy
	integrationToken, err := h.jwt.CreateIntegrationToken(clientID, workflowID)
	if err != nil {
		respondJSON(w, http.StatusInternalServerError, models.ErrorResponse{Detail: "Failed to create integration token"})
		return
	}

	// Filter headers
	headers := make(map[string]string)
	for k, v := range r.Header {
		lower := strings.ToLower(k)
		if !hopByHopHeaders[lower] && len(v) > 0 {
			headers[k] = v[0]
		}
	}

	// Set trusted T3Z headers
	headers["X-T3Z-Client-Id"] = clientID
	headers["X-T3Z-Workflow-Id"] = workflowID
	headers["X-T3Z-Integration-Token"] = integrationToken

	// Read and adapt body for Windmill parameters
	var bodyReader io.Reader = r.Body
	bodyBytes, err := io.ReadAll(r.Body)
	if err == nil && len(bodyBytes) > 0 {
		var rawObj map[string]interface{}
		if jsonErr := json.Unmarshal(bodyBytes, &rawObj); jsonErr == nil {
			if _, hasArgs := rawObj["args"]; !hasArgs {
				if _, hasPayload := rawObj["payload"]; !hasPayload {
					wrapped := map[string]interface{}{
						"args":    rawObj,
						"payload": rawObj,
					}
					for k, v := range rawObj {
						wrapped[k] = v
					}
					if wb, mErr := json.Marshal(wrapped); mErr == nil {
						bodyBytes = wb
					}
				}
			}
		}
		bodyReader = bytes.NewReader(bodyBytes)
	}

	// Forward to Windmill upstream
	resp, err := h.windmill.ProxyWebhook(targetPath, r.Method, r.URL.Query(), bodyReader, headers)
	if err != nil {
		respondJSON(w, http.StatusBadGateway, models.ErrorResponse{Detail: "Windmill upstream unavailable"})
		return
	}
	defer resp.Body.Close()

	// Copy response headers (excluding hop-by-hop)
	for k, v := range resp.Header {
		if !hopByHopHeaders[strings.ToLower(k)] && len(v) > 0 {
			w.Header().Set(k, v[0])
		}
	}

	w.WriteHeader(resp.StatusCode)
	_, _ = io.Copy(w, resp.Body)
}

func fmtPath(clientID, workflowID string) string {
	return "f/" + strings.ToLower(clientID) + "/" + workflowID
}
